source/proxy_strategy.c

```diff
@@ -1105,6 +1548,336 @@ struct aws_http_proxy_strategy_factory *aws_http_proxy_strategy_factory_new_tunn
+    void *internal_proxy_user_data) {
+
+    struct aws_http_proxy_strategy_tunneling_chain *ntlm_strategy = proxy_strategy->impl;
+    (void)ntlm_strategy;
+    (void)message;
+    (void)strategy_termination_callback;
+    (void)strategy_http_request_forward_callback;
+    (void)internal_proxy_user_data;
+
+    /*
+     * SA-TBI: modify message with ntlm auth data and call the request_forward callback or if
+     * encountering an error, invoke the strategy_termination callback.
+     *
+     * As written, a connection attempt using this strategy will hang because neither of these are currently
+     * invoked.
+     */
+    
+    /*Terminate when both states are not AWS_PSCS_READY*/
+    if ((ntlm_strategy->connect_state != AWS_PSCS_READY)
+        &&(ntlm_strategy->connect_sub_state != AWS_PSCS_SUB_STATE_READY)) {
+           strategy_termination_callback(message, AWS_ERROR_INVALID_STATE, internal_proxy_user_data);
+        return;
+    }
+
+    if (ntlm_strategy->connect_state == AWS_PSCS_READY) {
+
+        /* we first need to get the ntlm credential from user*/
+        char *ntlm_credential_user = get_ntlm_credential();
+        struct aws_byte_cursor ntlm_credential_tmp = aws_byte_cursor_from_c_str(ntlm_credential_user);
+        struct aws_string *ntlm_credential = aws_string_new_from_cursor(ntlm_strategy->allocator, &ntlm_credential_tmp);
+
+        /*transform the header with proxy authenticate:NTLM and NTLM credentials*/
+        if (s_add_ntlm_proxy_usertoken_authentication_header(ntlm_strategy->allocator, message, ntlm_credential)) {
+            strategy_termination_callback(message, aws_last_error(), internal_proxy_user_data);
+            return;
+        }
+    }
+    
+    else if (ntlm_strategy->connect_sub_state == AWS_PSCS_SUB_STATE_READY) {
+
+        char *ntlm_response_user = get_ntlm_response();
+        struct aws_byte_cursor ntlm_response_tmp = aws_byte_cursor_from_c_str(ntlm_response_user);
+        struct aws_string *ntlm_response =
+            aws_string_new_from_cursor(ntlm_strategy->allocator, &ntlm_response_tmp);
+
+        /*transform the header with proxy authenticate:NTLM and the response for challenge received from user*/
+        if (s_add_ntlm_proxy_usertoken_authentication_header(ntlm_strategy->allocator, message, ntlm_response)) {
+            strategy_termination_callback(message, aws_last_error(), internal_proxy_user_data);
+            return;
+        }
+
+        ntlm_strategy->connect_sub_state = AWS_PSCS_SUB_STATE_IN_PROGRESS;
+    }
+    
+    ntlm_strategy->connect_state = AWS_PSCS_IN_PROGRESS;
+    strategy_http_request_forward_callback(message, internal_proxy_user_data);
+}
+
+static int s_ntlm_on_incoming_header_adaptive(
+    struct aws_http_proxy_strategy *proxy_strategy,
+    enum aws_http_header_block header_block,
+    const struct aws_http_header *header_array,
+    size_t num_headers) {
+
+    struct aws_http_proxy_strategy_tunneling_chain *ntlm_strategy = proxy_strategy->impl;
```

Comment:
-chain, +ntlm

Reply:
Fixed, all strategy are now getting casted to correct structures